### PR TITLE
Include the newer Microsoft.PowerShell.ThreadJob module - fixes #19742

### LINKED ISSUE #19742: PowerShell ships ThreadJob not Microsoft.PowerShell.ThreadJob
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Why does PowerShell ship ThreadJob and not Microsoft.PowerShell.ThreadJob? The former hasn't been updated since 2019 while the latter has been updated in 2021.

@PaulHigin 

### Expected behavior

```console
Get-Module ThreadJob -ListAvailable

    Directory: C:\program files\powershell\7-preview\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Binary     2.0.3                 Microsoft.PowerShell.ThreadJob      Desk      Start-ThreadJob
```

### Actual behavior

```console
Get-Module ThreadJob -ListAvailable

    Directory: C:\program files\powershell\7-preview\Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Binary     2.0.3                 ThreadJob                           Desk      Start-ThreadJob
```

### Error details

_No response_

### Environment data

```powershell
7.4-preview3
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by kilasuit**
# PR Summary

Fixes #19742  by updating the package reference file for modules pulled from the gallery to also include the renamed `Microsoft.PowerShell.ThreadJob` module
 
## PR Checklist

- [ x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by kilasuit** [comment]
looking into the failures now

**Comment by ThomasNieto** [comment]
Doing a quick search on GitHub there are no references to the module qualified `ThreadJob\Start-ThreadJob`.

**Comment by kilasuit** [comment]
The test failure is one that I've seen come up before and has nothing to do with this PR `Set/New/Remove-Service cmdlet tests.Remove-Service can accept a ServiceController as pipeline input `

**Comment by kilasuit** [comment]
I think there will need to be a minor update to docs for this PR

**Comment by kasini3000** [comment]
this pr renamed Microsoft.PowerShell.ThreadJob module ?

Will this cause duplicate command names In multiple versions of pwsh ?
What mechanism is there to ensure that 【Microsoft. PowerShell. ThreadJob】 will be used before 【ThreadJob】when automatically importing module?

What are the differences between the two modules?
Why are the two modules forked? Why not update on ThreadJob?

File: tools/packaging/boms/mac.json [hunk 1]
```diff
@@ -347,6 +347,22 @@
...
+  {
+    "Pattern": "Modules/Microsoft.PowerShell.ThreadJob/*.dll",
+    "FileType": "NonProduct"
+  },
+  {
+    "Pattern": "Modules/Microsoft.PowerShell.ThreadJob/*.psd1",
+    "FileType": "NonProduct"
+  },
+  {
+    "Pattern": "Modules/Microsoft.PowerShell.ThreadJob/Microsoft.PowerShell.ThreadJob.pdb",
```
Comment (TravisEz13) [inline tools/packaging/boms/mac.json:359]: We should avoid shipping the pdb.  Can we file an issue in the threadjob repo

**Comment by adityapatwardhan** [comment]
Maintainers would like the Committee to review this breaking change of renaming the module from ThreadJob to Microsoft.PowerShell.ThreadJob.

File: tools/packaging/boms/mac.json (hunk 1 above)
Comment (kilasuit) [reply tools/packaging/boms/mac.json:359]: Issue filed in the [PowerShell/ThreadJob repo](https://github.com/PowerShell/ThreadJob/issues/28)
Looking into the repo and it's structure to see if can get a better understanding and potentially raise a PR myself for this

File: src/Modules/PSGalleryModules.csproj [hunk 2]
```diff
@@ -16,6 +16,7 @@
     <PackageReference Include="Microsoft.PowerShell.Archive" Version="1.2.5" />
     <PackageReference Include="PSReadLine" Version="2.2.6" />
     <PackageReference Include="ThreadJob" Version="2.0.3" />
+    <PackageReference Include="Microsoft.PowerShell.ThreadJob" Version="2.1.0" />
```
[the remark this answers is not in the document]
Comment (JosephColvin) [reply src/Modules/PSGalleryModules.csproj:19]: Maybe a Module Alias feature would help resolve this issue and others like it?

[the remark this answers is not in the document]
Comment (kilasuit) [reply src/Modules/PSGalleryModules.csproj:19]: @JosephColvin I've raised about exactly that possible enhancement in #19745 but that won't come in anytime soon I wouldn't think.

**Comment by kilasuit** [comment]
Currently waiting on this PR in the ThreadJob repo https://github.com/PowerShell/ThreadJob/pull/29 so can then push update to this PRT to resolve the conversation above by @adityapatwardhan

**Comment by daxian-dbw** [comment]
Still waiting on the committee to make a conclusion.

**Comment by JamesWTruher** [comment]
The @PowerShell/powershell-committee discussed this (6/28/23) and suggest that proper way to achieve this is as follows:
- Create a new version of the ThreadJob module which proxies to the new Microsoft.PowerShell.ThreadJob  module
  - the proxy should include a warning message that the ThreadJob module will be deprecated and removed in PS 7.6
- Include the new Microsoft.PowerShell.ThreadJob in the release

By taking step 1, we can avoid a breaking change and provide users a glide path to get off the old ThreadJob module which we would want to remove in PS7.6 (our next LTS after 7.4)

Implementation note:
  The release YAML will need updating for the Microsoft.PowerShell.ThreadJob module for proper authorization

A new issue should be opened to track the removal of the ThreadJob module in PS7.6.

**Comment by kilasuit** [comment]
Just so that I am clear from the above

Currently we are waiting on a new ThreadJob release -> then can update this PR with that version number & inc the Microsoft.PowerShell.ThreadJob as currently proposed  & all's good for the time being?

**Comment by JamesWTruher** [comment]
> @JamesWTruher - Just so that I am clear from the above
> 
> Currently we are waiting on a new ThreadJob release -> then can update this PR with that version number & inc the Microsoft.PowerShell.ThreadJob as currently proposed & all's good for the time being?

sort of.

We want to be sure that the current ThreadJob module is updated with a deprecation notice (and create a proxy from `ThreadJob\Start-ThreadJob`  to the `Microsoft.PowerShell.ThreadJob\Start-ThreadJob` cmdlet. 

But essentially, we will ship the old `ThreadJob` module _and_ the new `Microsoft.PowerShell.ThreadJob` module. If you must have the old behavior, you would be able to achieve this via: `Import-Module -RequiredVersion x -Name ThreadJob` but just typing `Start-ThreadJob` will auto-load the _new_ `Microsoft.PowerShell.ThreadJob` module (which is why this is still technically a breaking change).

explicitly importing the new `ThreadJob` module should result in a warning message with the "this is going away, so you should use the Microsoft.PowerShell.ThreadJob module"

**Comment by kilasuit** [comment]
Thanks for clarifying @JamesWTruher that was how I understood it from your comment 🙂

**Comment by daxian-dbw** [comment]
@JamesWTruher A few questions:

> but just typing `Start-ThreadJob` will auto-load the new `Microsoft.PowerShell.ThreadJob` module

1. How would we guarantee `Microsoft.PowerShell.ThreadJob` will be the one loaded when just typing `Start-ThreadJob`?

2. Who will be working on the proxy ThreadJob module? Is this PR blocked until the new proxy `ThreadJob` module is released? If so, I would rather to have this PR closed for now.

**Comment by JamesWTruher** [comment]
> @JamesWTruher A few questions:
> 
> > but just typing `Start-ThreadJob` will auto-load the new `Microsoft.PowerShell.ThreadJob` module
> 
> 1. How would we guarantee `Microsoft.PowerShell.ThreadJob` will be the one loaded when just typing `Start-ThreadJob`?
> 2. Who will be working on the proxy ThreadJob module? Is this PR blocked until the new proxy `ThreadJob` module is released? If so, I would rather to have this PR closed for now given that we don't have an ETA.

1. Currently, autoloading loads the first module _found_, and `Microsoft.PowerShell.ThreadJob` will be found before `ThreadJob`
2. @SteveL-MSFT will have a more informed answer for this

**Comment by SteveL-MSFT** [comment]
Yes, this PR is blocked until someone can develop the ThreadJob proxy module (could be part of this PR as it's just a script module).

**Comment by daxian-dbw** [comment]
The proxy module shouldn't be developed in this PR, as the code should be in a different repo.
If this PR is blocked on that, then I will close this PR for now. Does it sound good to both of you?

**Comment by JosephColvin** [comment]
James,
 PowerShell will auto load Microsoft.PowerShell.ThreadJob because of the alphabetical order being above ThreadJob (M > T), correct?

Joseph Colvin
IT Coordinator
Warwick Public School<https://www.warwick.k12.nd.us/>
School Phone: 701-294-2561
My Cell: 701-219-4044
My Office Phone: 701-294-2561 Ext. 328
“So you never know who you touch. You never know how or when you'll have an impact, or how important your example can be to someone else" ― Denzel Washington, A Hand to Guide Me

"When men speak ill of you, live so as nobody may believe them." ― Plato

“time, man....time is an illusion. The river flows at an uneven rate but we all drift along in it without noticing until we look back.” – John Gieser

 Save a tree, think before you print.

***@***.***?anonymous&ep=signature>                Book time to meet with ***@***.***?anonymous&ep=signature>

________________________________

**Comment by SteveL-MSFT** [comment]
> The proxy module shouldn't be developed in this PR, as the code should be in a different repo. If this PR is blocked on that, then I will close this PR for now. Does it sound good to both of you?

The proxy module is really only relevant as part of PS7 due to the breaking change of ThreadJob being included in PS7 prior to renaming to Microsoft.PowerShell.ThreadJob.  It's really only to provide compat for one (or two) release and provide a warning message then we can remove it.

**Comment by daxian-dbw** [comment]
But `ThreadJob` is a published module in PowerShell Gallery (v2.0.3 now). If we are going to have a proxy module for it, we'd better release it to PowerShell Gallery too, maybe as v2.1, so as formally deliver the message that this module is deprecating and `Microsoft.PowerShell.ThreadJob` should be one to use going forward.

Once the proxy module is published, PowerShell just need to pull it during the build, and then it's time to include `Microsoft.PowerShell.ThreadJob` module.

**Comment by SteveL-MSFT** [comment]
@daxian-dbw I agree it would make sense to have the ThreadJob module on PSGallery also indicate being deprecated and using Microsoft.PowerShell.ThreadJob instead.  In that case, I would also agree that proxy module would be done outside of this PR and repo since it's published to PSGallery.  It would make sense to have that proxy developed in https://github.com/powershell/threadjob

**Comment by daxian-dbw** [comment]
Opened https://github.com/PowerShell/ThreadJob/issues/30 to track the proxy module work. I will close this PR for now.

**Pull request closed without merging**